rust/src/pool/mod.rs

```diff
@@ -0,0 +1,409 @@
+            value: 0,
+            children: vec![],
+        };
+        let root_id = pool.set(root).unwrap();
+        let root = pool.get(root_id).unwrap();
+        populate_tree(&mut root.write().unwrap(), 0, TREE_DEPTH, 0, &pool.clone());
+
+        let mut cases = vec![];
+        generate_cases_recursive(&mut cases, &mut vec![], TREE_DEPTH);
+        // IMO this is clearer than filtering the elements in the for loop below.
+        #[allow(clippy::needless_collect)]
+        let paths: Vec<Vec<u32>> = cases.into_iter().map(|c| c.path).collect();
+
+        let mut handles = vec![];
+        for (i, path) in paths.into_iter().enumerate() {
+            // Spawn set
+            handles.push(thread::spawn({
+                let pool = pool.clone();
+                let root = pool.get(root_id).unwrap();
+                let path = path.clone();
+                move || {
+                    set_value_at_path(&root.read().unwrap(), &path, i as u32, &pool.clone());
+                }
+            }));
+            // Spawn get
+            handles.push(thread::spawn({
+                let pool = pool.clone();
+                let root = pool.get(root_id).unwrap();
+                move || {
+                    let _ = read_from_tree_path(&root.read().unwrap(), &path, &pool);
+                }
+            }));
+        }
+
+        for h in handles {
+            h.join().unwrap();
+        }
+    }
+
+    /// A simple in-memory pool of nodes for testing purposes.
+    struct FakeNodePool {
+        nodes: Mutex<HashMap<u32, Arc<RwLock<Node>>>>,
+    }
+
+    impl Pool for FakeNodePool {
+        type Key = u32;
+        type Type = Node;
+
+        fn get(
+            &self,
+            id: Self::Key,
+        ) -> Result<
+            PoolEntry<impl DerefMut<Target = Self::Type> + Send + Sync + 'static>,
+            crate::error::Error,
+        > {
+            self.nodes
+                .lock()
+                .unwrap()
+                .get(&id)
+                .cloned()
+                .map(super::PoolEntry::new)
+                .ok_or(Error::Storage(storage::Error::NotFound))
+        }
+
+        fn set(&self, value: Node) -> Result<Self::Key, crate::error::Error> {
+            let node = Arc::new(RwLock::new(value));
+            let mut nodes = self.nodes.lock().unwrap();
+            let id = nodes.len() as u32 + 1;
+            nodes.insert(id, node);
+            Ok(id)
+        }
+
+        fn delete(&self, id: Self::Key) -> Result<(), crate::error::Error> {
+            self.nodes
+                .lock()
+                .unwrap()
+                .remove(&id)
+                .map(|_| ())
+                .ok_or(Error::Storage(storage::Error::NotFound))
+        }
+
+        fn flush(&self) -> Result<(), crate::error::Error> {
+            Ok(())
+        }
+    }
+
+    /// A simple tree node for testing purposes.
+    /// It stores a u32 payload a list of children node IDs.
+    struct Node {
```

Comment:
```suggestion
    struct TestNode {
```

Why not have this at the start of the module? Now I have to scroll down to figure out what this type is while reading the tests.